tapdb/multiverse.go

```diff
@@ -64,21 +73,420 @@ type BatchedMultiverse interface {
+	rootIndex := r.rootIndex.Load()
+
+	root, ok := rootIndex.Load(treeID(id.String()))
+	if ok {
+		r.Hit()
+		return root
+	}
+
+	r.Miss()
+
+	return nil
+}
+
+// cacheRoot stores the given root in the cache.
+func (r *rootNodeCache) cacheRoot(id universe.Identifier,
+	root universe.Root) {
+
+	rootIndex := r.rootIndex.Load()
+	rootIndex.Store(treeID(id.String()), &root)
+}
+
+// cacheRoots stores the given roots in the cache.
+func (r *rootNodeCache) cacheRoots(q universe.RootNodesQuery,
+	rootNodes []universe.Root) {
+
+	log.Debugf("caching num_roots=%v", len(rootNodes))
+
+	// Store the main root pointer, then update the root index.
+	rootPageCache := r.allRoots.Load()
+	_, _ = rootPageCache.Put(newRootPageQuery(q), rootNodes)
+
+	rootIndex := r.rootIndex.Load()
+	for _, rootNode := range rootNodes {
+		rootNode := rootNode
+
+		idStr := treeID(rootNode.ID.String())
+		rootIndex.Store(idStr, &rootNode)
+	}
+}
+
+// wipeCache wipes all the cached roots.
+func (r *rootNodeCache) wipeCache() {
+	log.Debugf("wiping universe cache")
+
+	rootCache := lru.NewCache[rootPageQuery, universeRootPage](
+		numCachedProofs,
+	)
+	r.allRoots.Store(rootCache)
+
+	r.rootIndex.Store(&rootIndex{})
+}
+
+// cachedLeafKeys is used to cache the set of leaf keys for a given universe.
+//
+// TODO(roasbeef); cacheable[T]
+type cachedLeafKeys []universe.LeafKey
+
+// Size just returns 1, as we cache based on the total number of assets, but
+// not the sum of their leaves.
+func (c cachedLeafKeys) Size() (uint64, error) {
+	return uint64(1), nil
+}
+
+// numMaxCachedPages is the maximum number of pages we'll cache for a given
+// page cache. Each page is 512 items, so we'll cache 10 of them, up to 5,120
+// for a given namespace.
+const numMaxCachedPages = 1000
+
+// leafQuery is a wrapper around the existing UniverseLeafKeysQuery struct that
+// doesn't include a pointer so it can be safely used as a map key.
+type leafQuery struct {
+	sortDirection universe.SortDirection
+	offset        int32
+	limit         int32
+}
+
+// newLeafQuery creates a new leaf query.
+func newLeafQuery(q universe.UniverseLeafKeysQuery) leafQuery {
+	return leafQuery{
+		sortDirection: q.SortDirection,
+		offset:        int32(q.Offset),
+		limit:         int32(q.Limit),
+	}
+}
+
+// leafPageCache caches the various paginated responses for a given treeID.
+type leafPageCache struct {
+	*lru.Cache[leafQuery, *cachedLeafKeys]
+}
+
+// Size returns the number of elements in the leaf page cache.
+func (l *leafPageCache) Size() (uint64, error) {
+	return uint64(l.Len()), nil
+}
+
+// leafKeysCache is used to cache the set of leaf keys for a given universe.
+// For each treeID we store an inner cache for the paginated responses.
+type leafKeysCache = lru.Cache[treeID, *leafPageCache]
+
+// universeLeafCaches is used to cache the set of leaf keys for a given
+// universe.
+type universeLeafCache struct {
+	sync.Mutex
+
+	leafCache *leafKeysCache
```

Comment:
IIUC there is no limit on how many leaf page caches we make, but one limit on elements shared over all leaf page caches?

Reply:
So there's a limit on the total amount of active caches for a given universe (`treeID`), this is set to 50k rn. Then for each of chose, due to the pagination, we'll cache different pages (up to 1k total pages). I think we can tune these values further, once most clients are upgraded, that way the LRU will actually cache the most relevant items, rather than needing to cache it all.